src/tiled/tileseteditor.cpp

```diff
@@ -435,8 +436,41 @@ void TilesetEditor::addTiles()
         QPixmap image;
     };
     QVector<LoadedFile> loadedFiles;
-
+    // If the tile is already in the tileset, warn user and confirm addition
+    bool dontAskAgain = false, rememberOption = true;
     for (const QString &file : files) {
+        bool addImage = true;
+        const QMap<int, Tile*> addedTiles = mCurrentTilesetDocument->tileset().data()->tiles();
+        for (auto tile : addedTiles) {
+            if (tile->imageSource() == file) {
+                if (dontAskAgain) {
+                    addImage = rememberOption;
+                    break;
+                }
+                QCheckBox *checkBox = new QCheckBox(tr("Apply this action to all tiles"));
+                QMessageBox warning(QMessageBox::Warning,
+                            tr("Add Tiles"),
+                            tr("Tile \"%1\" already exists in the tileset!").arg(file),
+                            QMessageBox::Yes | QMessageBox::No,
+                            mMainWindow->window());
+                warning.setDefaultButton(QMessageBox::Yes);
+                warning.setInformativeText(tr("Add anyway?"));
+                warning.setCheckBox(checkBox);
+                if (warning.exec() != QMessageBox::Yes) {
+                    rememberOption = addImage = false;
+                }
+                else {
+                    rememberOption = true;
+                }
+                if (checkBox->checkState() == Qt::Checked) {
+                    dontAskAgain = true;
+                }
+                break;
```

Comment:
Although it would avoid the nested loop, it could also complicate the logic, since the use of the `dontAskAgain` variable would only be possible if as a reference parameter in the separated method, or if it become a class variable of tileseteditor. This happens because the `QMessageBox` solves two problems in the same place (should add tile and should remember this option). So for me the cleaner way of implementing this separately would be a boolean method that returns if the tile should be added plus a reference parameter to set if that action should be remembered, if that is what you meant I can make the changes, but I still think that as a matter of clarity, this code is easier to understand.